fiaas_deploy_daemon/extension_hook_caller.py

```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env python
+# -*- coding: utf-8
+
+# Copyright 2017-2021 The FIAAS Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import json
+import logging
+from collections import OrderedDict
+
+from six import string_types
+
+
+LOG = logging.getLogger(__name__)
+
+
+class ExtensionHookCaller(object):
+
+    def __init__(self, config, session):
+        self._url = config.hook_service
+        self._session = session
+
+    def apply(self, kind, obj, app_spec):
+        if self._url is None:
+            return obj
+        url = str(self._url) + "/fiaas/deploy/" + str(kind)
+        dump = json.dumps({"object": obj.as_dict(), "application": self._app_spec_to_dict(app_spec)})
```

Comment:
I think it's worth considering Data on the inside vs Data on the outside (https://queue.acm.org/detail.cfm?id=3415014) here, and `app_spec` is definitely on the inside. I think it would be better if the object posted to the hook was the original `Application` object (or the `spec` from that), but since we generally don't pass that around (since it's an "outside object"), I can see how it would be difficult to use that here.

Tying the hook service to our internal representation (the `app_spec`) limits our ability to evolve the `app_spec` object to our needs, because it might cause problems for the hook service.

Would it make sense to attach the original object to the app_spec, so that it is available for reference in cases like this one?

Reply:
Yes, it makes totally sense :)